posthog/special_migrations/setup.py

```diff
@@ -0,0 +1,110 @@
+from semantic_version.base import Version
+
+from posthog.settings import AUTO_START_SPECIAL_MIGRATIONS, DEBUG, E2E_TESTING, SKIP_SERVICE_VERSION_REQUIREMENTS, TEST
+from posthog.special_migrations.definition import SpecialMigrationDefinition
+from posthog.utils import print_warning
+from posthog.version import VERSION
+
+ALL_SPECIAL_MIGRATIONS: Dict[str, SpecialMigrationDefinition] = {}
+
+SPECIAL_MIGRATION_TO_DEPENDENCY: Dict[str, Optional[str]] = {}
+
+# inverted mapping of SPECIAL_MIGRATION_TO_DEPENDENCY
+DEPENDENCY_TO_SPECIAL_MIGRATION: Dict[Optional[str], str] = {}
+
+
+POSTHOG_VERSION = Version(VERSION)
+
+SPECIAL_MIGRATIONS_MODULE_PATH = "posthog.special_migrations.migrations"
+SPECIAL_MIGRATIONS_EXAMPLE_MODULE_PATH = "posthog.special_migrations.examples"
+
+
+def setup_special_migrations():
+    """
+    Execute the necessary setup for special migrations to work:
+    1. Import all the migration definitions 
+    2. Create a database record for each
+    3. Check if all migrations necessary for this PostHog version have completed (else don't start)
+    4. Populate a dependencies map and in-memory record of migration definitions
+    """
+
+    from posthog.models.special_migration import SpecialMigration, get_all_completed_special_migrations
+
+    if TEST or E2E_TESTING or SKIP_SERVICE_VERSION_REQUIREMENTS:
+        print_warning(["Skipping special migrations setup. This is unsafe in production!"])
+        return
+
+    all_migrations = import_submodules(SPECIAL_MIGRATIONS_MODULE_PATH)
+
+    if DEBUG:
+        all_migrations["example"] = import_submodules(SPECIAL_MIGRATIONS_EXAMPLE_MODULE_PATH)["example"]
+
+    for name, module in all_migrations.items():
+        ALL_SPECIAL_MIGRATIONS[name] = module.Migration()
+
+    applied_migrations = set(instance.name for instance in get_all_completed_special_migrations())
+    unapplied_migrations = set(ALL_SPECIAL_MIGRATIONS.keys()) - applied_migrations
+
+    first_migration = None
+    for migration_name, migration in ALL_SPECIAL_MIGRATIONS.items():
+
+        SpecialMigration.objects.get_or_create(
+            name=migration_name,
+            description=migration.description,
+            posthog_max_version=migration.posthog_max_version,
+            posthog_min_version=migration.posthog_min_version,
+        )
+
+        dependency = migration.depends_on
+
+        if not dependency:
+            if first_migration:
+                raise ImproperlyConfigured(
+                    "Two or more special migrations have no dependency. Make sure only the first migration has no dependency."
+                )
+
+            first_migration = migration_name
+
+        SPECIAL_MIGRATION_TO_DEPENDENCY[migration_name] = dependency
+
+        if migration_name in unapplied_migrations and POSTHOG_VERSION > Version(migration.posthog_max_version):
+            raise ImproperlyConfigured(f"Migration {migration_name} is required for PostHog versions above {VERSION}.")
```

Comment:
Important: 
1. Help users know how to recover faster: We should crash here, but let's provide the user with better info, e.g. let's tell them explicitly what version they need to update back to for a better experience (potentially link to a doc page explaining what special migrations are & how to check from db what version to go back to).
2. This shouldn't be the first/only place we crash: We should crash earlier during the update path, specifically events/web should never get updated if the required special migrations aren't done, instead we should fail a pre-hook. That way the upgrade fails and their instance is still in a good state (as opposed to being down for ingestion! and not knowing what to do exactly, which will make support not fun). Specifically I'm suggesting we create a pre-hook job (similar to migrations job and ran after the migrations job), that checks that all the necessary special migrations are done or we are on a fresh install.

Reply:
The problem is the `helm upgrade` will succeed and consider everything done (for sure if we don't use `--atomic` and maybe even then) & then we're down for ingestion as the events pod will be down. 

Thinking about this further we actually need to run the check before doing normal migrations (if we want to assume normal migrations added after posthog version upper bound are safe to use, which we should as that makes everything else simple). Hence a proposal = add a pre-hook before migration job, that 
1. does nothing if it's a `helm install` (which means we'll run all migrations and then all special migrations)
2. for `helm upgrade` it checks the special migrations DB state against the version we're about to upgrade to and complains/fails if we don't have all special migrations ran yet that need to be completed for that version. (this way we are guaranteed to not run normal migrations that were added after the upper bound special migrations version).

Reply:
I believe I've solved this problem - any Django command, be it to run the server or migrate would go through the setup code path. 

This was actually problematic because a new deploy would no start and get a warning about dependencies necessary for this version. 

What I've done is added a command to run special migrations. What it will do is try to run all migrations we're above the range on if their `is_required` check returns true. If one of them can't run, we'll crash.

That means:

1. Users on fresh deploys will get all necessary migrations executed synchronously before starting up
2. For users updating we will also get migrations they didn't apply that are necessary for the upcoming version and try to run them synchronously. If we manage, great. Else they will run into the timeout, and their instance will crash in that way.